Firestore/Source/Local/FSTLevelDBMigrations.mm

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/Source/Local/FSTLevelDBMigrations.h"
+
+#include <leveldb/db.h>
+#include <leveldb/write_batch.h>
+
+#import "Firestore/Protos/objc/firestore/local/Target.pbobjc.h"
+#import "Firestore/Source/Local/FSTLevelDBKey.h"
+#import "Firestore/Source/Local/FSTLevelDBUtil.h"
+#import "Firestore/Source/Util/FSTAssert.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+using leveldb::DB;
+using leveldb::Status;
+using leveldb::Slice;
+using leveldb::WriteOptions;
+
+static void EnsureTargetGlobal(std::shared_ptr<DB> db) {
+  FSTPBTargetGlobal *targetGlobal = [FSTLevelDBUtil readTargetMetadataFromDB:db];
+  if (!targetGlobal) {
+    targetGlobal = [FSTPBTargetGlobal message];
+    NSData *data = [targetGlobal data];
+    Slice value((const char *)data.bytes, data.length);
+    Status status = db->Put(WriteOptions(), [FSTLevelDBTargetGlobalKey key], value);
+    if (!status.ok()) {
+      FSTCFail(@"Failed to save  metadata: %s", status.ToString().c_str());
+    }
+  }
+}
+
+static void SaveVersion(std::shared_ptr<DB> db, FSTLevelDBSchemaVersion version) {
+  std::string key = [FSTLevelDBVersionKey key];
+  std::string version_string = std::to_string(version);
+  Status status = db->Put(WriteOptions(), key, version_string);
+  if (!status.ok()) {
+    FSTCFail(@"Saving schema version failed with status: %s", status.ToString().c_str());
+  }
+}
+
+@implementation FSTLevelDBMigrations
+
++ (FSTLevelDBSchemaVersion)schemaVersionForDB:(std::shared_ptr<DB>)db {
+  std::string key = [FSTLevelDBVersionKey key];
+  std::string version_string;
+  Status status = db->Get([FSTLevelDBUtil standardReadOptions], key, &version_string);
+  if (status.IsNotFound()) {
+    return 0;
+  } else {
+    return stoi(version_string);
+  }
+}
+
++ (void)runMigrationsToVersion:(FSTLevelDBSchemaVersion)version onDB:(std::shared_ptr<DB>)db {
+  FSTLevelDBSchemaVersion currentVersion = [self schemaVersionForDB:db];
+  switch (currentVersion) {
+    case 0:
+      EnsureTargetGlobal(db);
```

Comment:
Performing migrations outside a write group makes me nervous because it requires every write to be idempotent. It has the potential to complicate migrations that potentially write more than one row.

If we can I think we'd be better off passing a write group into each function like EnsureTargetGlobal and then committing either once per migration version (including a bump to that version) or once overall.

This would also have the side-effect of simplifying the migrations themselves since each one no longer has to deal with committing. For example, EnsureTargetGlobal is cut down to this:

```c++
static void EnsureTargetGlobal(std::shared_ptr<DB> db, FSTWriteGroup *group) {
  FSTPBTargetGlobal *targetGlobal = [FSTLevelDBUtil readTargetMetadataFromDB:db];
  if (!targetGlobal) {
    [group setMessage:[FSTPBTargetGlobal message] forKey:[FSTLevelDBTargetGlobalKey key]];
  }
}
```

Reply:
Done.